Implement limit offset with optimized tx-add map

coderofstuff opened the pull request:
Add the endpoint that returns all transactions within a given limit and offset

New endpoint: /address/{address}/transctions/v2 (we can replace this or rename this however we want)

This combines the old functionality of /address/{address}/transactions and /transactions/search and includes usage of the optimization table `tx_id_address_mapping`.

Depends on:
- `tx_id_address_mapping` existing and fully filled

Notes:
- Does not include `is_accepted` filtering yet

coderofstuff committed:
Implement limit offset with optimized tx-add map

Add the endpoint that returns all transactions within a given limit and offset

lAmeR1 commented on endpoints/get_address_transactions.py:
```diff
@@ -60,3 +63,42 @@ async def get_transactions_for_address(
+        limit: int = Query(
+            description="The number of records to get",
+            ge=1,
+            le=500,
```
perfect, had the same numbers in my mind

lAmeR1 commented on endpoints/get_address_transactions.py:
```diff
@@ -60,3 +63,42 @@ async def get_transactions_for_address(
         "transactions": tx_list
     }
+
+@app.get("/addresses/{kaspaAddress}/transactions/v2",
```
should we create a version at the beginning ?

coderofstuff replied:
You mean `/v2/addresses/{kaspaAddress}/transactions`?

I can also simply change the endpoint to `/addresses/{kaspaAddress}/transactions-full` to avoid versioning

coderofstuff replied:
I used `/addresses/{kaspaAddress}/full-transactions` at 27235208739ff1f1cb6ec9537a57fce8419e6d0e

lAmeR1 commented on endpoints/get_address_transactions.py:
```diff
@@ -60,3 +63,42 @@ async def get_transactions_for_address(
+    async with async_session() as s:
+        # Doing it this way as opposed to adding it directly in the IN clause
+        # so I can re-use the same result in tx_list, TxInput and TxOutput
+        txWithinLimitOffset = await s.execute(select(TxAddrMapping.transaction_id)
```
If you don't mind use lowercase with underscores. Trying to follow PEP8

https://peps.python.org/pep-0008/#function-and-variable-names

coderofstuff replied:
Resolved with ff606f4

lAmeR1 commented on endpoints/get_address_transactions.py:
```diff
@@ -60,3 +63,42 @@ async def get_transactions_for_address(
+            .filter(TxAddrMapping.address == kaspaAddress)
+            .limit(limit)
+            .offset(offset)
+            .order_by(TxAddrMapping.block_time.desc())
```
Let's query/join the 'transactions' here I guess without using the search function. Will be fast - no?

lAmeR1 replied:
SELECT * FROM 
  transactions,
(SELECT * FROM tx_id_address_mapping WHERE address='kaspa:-....'
ORDER by block_time DESC
LIMIT 50
OFFSET 10000) as subquery

WHERE transactions.transaction_id = subquery.transaction_id

coderofstuff replied:
I used the function that exists in search because that one actually performs 3 queries: one for the TxModel and, depending on if `fields` contains `inputs` or `outputs` it will query those too.

If we only did the one query in `search` then I would agree that we would simply do the subquery like this.

However, if we don't really need the tx_output query and tx_input query that was being done in `search` I can make this endpoint do the query itself.

lAmeR1 commented:
filling the table still.. expect around 2 more days for filling the inputs

coderofstuff committed:
Follow variable naming conventions

coderofstuff committed:
Rename endpoint and fix tags

lAmeR1 commented on endpoints/get_address_transactions.py:
```diff
@@ -60,3 +63,42 @@ async def get_transactions_for_address(
+
+        tx_ids_in_page = [x[0] for x in tx_within_limit_offset.all()]
+
+    return await search_for_transactions(TxSearch(transactionIds=tx_ids_in_page), fields)
```
need to add `.order_by(Transaction.block_time.desc()))` into search_for_transactions so the ORDER is guaranteed

coderofstuff replied:
Resolved in c78c378

coderofstuff committed:
Guarantee search transaction order

Add `order_by` to ensure the transactions are sorted in descending order
for the results of `search_for_transactions`
